### ISSUE TITLE
Warn of obsucure floating-point literals

### ISSUE THREAD

**Issue by tamaroning** [rust-lang/rust-clippy#8424]
### What it does

This lint warns of misleading floating-point literals.
In the example, a has type i32 because floating-point literals must starts with a decimal literal.
Instead of suffixes, we should use type ascriptions.

### Lint Name

obscure_float_literal

### Category

style

### Advantage

- Avoid ambiguity

### Drawbacks

none

### Example

```rust
let a = 0x1_f32; // == 0x1F32
```

Could be written as:

```rust
let a = 0x1 as f32;
```

**Comment by xFrednet** [comment]
This can be a lint. However, it should be in the `restriction` or some other allow-by-default group, as this is very debatable. We have some lints like [`default_numeric_fallback`](https://rust-lang.github.io/rust-clippy/master/index.html#default_numeric_fallback) that suggest using the type suffixes. I also prefer having the type as a suffix :upside_down_face:

**Comment by 5225225** [comment]
> I also prefer having the type as a suffix

The issue is that the `_f32` here is *not* a type suffix.

It's literally just `0x1f32` which is a perfectly valid hex literal for an i32. You *can't* denote a hex literal to be a float.

So this is *incredibly* misleading code, and should probably be in `suspicious`, if not `correctness` (if the user writes `_f32` or `_f64`)

**Comment by xFrednet** [comment]
Ahh yes, I misunderstood the issue. Thank you very much for pointing that out! Then it should be in `suspicious` or `correctness` as you suggested. :upside_down_face:

**Comment by stanislav-tkach** [comment]
I found the following case even more surprising: `0x1_f32`.
